### PR TITLE
Dynamic Linux Slaves

### PR DESCRIPTION
**Pull request by jacderida**
Hi Stephen/Calum,

This is for dynamic slaves that are spun up with the EC2 plugin. For now it is limited to Linux. As I discussed with Stephen, the situation for getting it to work with Windows is a bit more complicated and requires a day or 2 dedicated to that, so we agreed to go with a static instance for now.

This time you'll want to kick off a build after you've spun up the infrastructure, so you can see the Linux slaves spin up dynamically. Note that before you start the build you won't see any Linux slaves - that's expected; you'll see a Windows slave that should be connected, and the Mac slave will be marked as offline, as expected for now. The build won't be able to pass the first stage due to the Mac not being available. That's expected. The Windows build will fail because of the missing `mingw` package (it won't be able to find `gcc`).

Some important new instructions have been added to the [README](https://github.com/jacderida/safe-build-infrastructure/tree/dynamic_slaves):

* The issue with the dollar in the Windows password is gone. I just wrapped the Ansible call in a shell script, so that should deal with the dollar OK.
* Before kicking off the infrastructure provisioning from the Bastion, you need to provide the ID of the private subnet for the slaves. The instructions in the README have been updated for this.
* Under 'Configure Jenkins' there is one additional manual step for providing an SSH key for the slaves. It will be possible to automate this, but the key needs to be provided in a very specific format and I didn't want to spend any time on that just now.

When you have the infrastructure up and running, create the pipeline for SCL using the instructions under 'Configure Jenkins', then run it. After a couple of minutes or so you should see 2 Linux slaves appear on the main page in the GUI. At first they will have a red cross indicating Jenkins can't establish an SSH connection, but that's just because it's waiting for the sshd service on the machine to come online. If you don't forget the manual step to provide the private key, they should connect after a little while.

Cheers,

Chris

### Windows Machine Provisioning

This applies to the `make provision-prod-jenkins-environment-aws` part of the process.

A couple of things to look out for:

* There was one occasion this afternoon where I saw the provision for the Windows machine fail sporadically when trying to install the `awscli` package. If it fails to install any packages, just run the provisioning again. The process is idempotent, so you can simply run `make provision-prod-jenkins-environment-aws` to do it again.
* If you see Ansible hang during the 'run get agent secret key script' task, this is most likely because you've provided the wrong hostname for Jenkins (as I did once this morning). Stop the process with `ctrl+c` and then make sure you have the correct IP address for `JENKINS_MASTER_HOSTNAME`.

### Windows SCL Build

If you didn't restart the Windows instance after provisioning, the Windows build steps in the pipeline will fail instantly complaining it can't find Cargo.

### Dynamic Slaves

If they don't show up or end up connecting, you can identify the problem by looking at the Jenkins log. It's likely there was possibly a problem with some of the manual steps.

To see the Jenkins logs:

* SSH to the Bastion host the same way you do for provisioning the infrastructure
* SSH from there to the Jenkins host (get the private IP from the AWS GUI): `ssh ansible@<private ip of jenkins master>`
* Get the ID of the Jenkins container: `sudo docker ps -a`
* Tail the logs: `sudo docker logs -f <jenkins container id>`

If the problem isn't obvious from the log, it'll probably need to wait till I come back.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -119,6 +117,9 @@ Now perform the following steps (all of these have to be applied to the Bastion
 * Set your AWS secret key: `export AWS_SECRET_ACCESS_KEY=<secret access key>`
 * Get the Jenkins IP address from the AWS GUI and set that: `export JENKINS_MASTER_HOSTNAME=<ip address of jenkins master>`
 * Get the Windows password from the AWS GUI and set that: `export JENKINS_WINDOWS_SLAVE_PASSWORD='<windows password>'` (note the password must go inside single quotes to prevent Bash from interpreting special characters)
+* Get the ID of the private subnet for the slaves and set that:
+    - In the AWS GUI go to Services -> VPC -> Subnets -> jenkins_environment-private-eu-west-2 and copy the ID of the subnet
+    - `export SLAVE_SUBNET_ID='<private subnet ID>'`
```
Comment (calumcraig) [inline README.md]: Need to remove the single quotes around `<private subnet ID>`

**Review by calumcraig — CHANGES_REQUESTED** [review]
I've commented on line 122 of README

**Review by calumcraig — COMMENTED** [review]
Have successfully made the build environment and provisioned Jenkins as described in this PR. After kicking off an SCL build job in Jenkins, two Linux slaves did spin up dynamically and come online after some time, resulting in the final configuration:
image

**Comment by calumcraig** [comment]
Upon trying to tear down this set up using `cd terraform/prod && terraform destroy` I had some errors:
```
2 error(s) occurred:

* aws_security_group.linux_slaves (destroy): 1 error(s) occurred:

* aws_security_group.linux_slaves: DependencyViolation: resource sg-0b0e8e8ee2b54f843 has a dependent object
	status code: 400, request id: e993c1e7-3265-459e-a3dc-761659cbc8f0
* module.vpc.aws_subnet.private (destroy): 1 error(s) occurred:

* aws_subnet.private: Error deleting subnet: timeout while waiting for state to become 'destroyed' (last state: 'pending', timeout: 10m0s)
```
And the two slaves were still seen as running in the AWS GUI. I was able to terminate them from the GUI however.

File: README.md (hunk 1 above)
Comment (S-Coyle) [reply README.md]: Note that on macOS I tried this with the single quotes (') included around the subnet ID and it worked fine

**Review by S-Coyle — DISMISSED** [review]
✅ Was able to follow the instructions and see the Linux slaves spin up when I kicked off a SCL job.
✅ Cancelled the running jobs & waited 30 minutes to confirm that the Linux slaves were torn down automatically.
✅ Once both Linux slaves were torn down I was able to run `terraform destroy` and successfully remove the remaining infrastructure.

👍

**Review by S-Coyle — DISMISSED** [review]
Also now tested synced folders fix on macOS - works as expected.

File: README.md (hunk 1 above)
Comment (jacderida) [reply README.md]: I'll remove this just for consistency anyway, since it was called out for the password because that's strictly non-optional.

**Review by S-Coyle — APPROVED** [review]
Approved 👍 
Note that I've not tried exporting the subnet ID without the single quotation marks, as per the most recent commits added, but can't see this being a problem as there shouldn't be special characters in this variable

**Review by calumcraig — APPROVED** [review]

**Pull request merged**